File: src/main/worktree.ts
import { execFile } from 'child_process'
import { promisify } from 'util'
import { promises as fs } from 'fs'
import { homedir } from 'os'
import { join, basename } from 'path'

const exec = promisify(execFile)

const WORKTREE_ROOT = join(homedir(), '.operator', 'worktrees')

async function git(cwd: string, ...args: string[]): Promise<string> {
  const { stdout } = await exec('git', args, { cwd, maxBuffer: 1024 * 1024 })
  return stdout.trim()
}

export interface RepoInfo {
  isRepo: boolean
  /** Top level of the working tree (only present when isRepo). */
  root?: string
  /** Currently checked-out branch (only present when isRepo and on a branch). */
  branch?: string
}

export async function inspectRepo(cwd: string): Promise<RepoInfo> {
  try {
    const root = await git(cwd, 'rev-parse', '--show-toplevel')
    let branch: string | undefined
    try {
      const b = await git(cwd, 'rev-parse', '--abbrev-ref', 'HEAD')
      if (b && b !== 'HEAD') branch = b
    } catch { /* detached or empty repo */ }
    return { isRepo: true, root, branch }
  } catch {
    return { isRepo: false }
  }
}

export interface WorktreeCreateResult {
  path: string
  branch: string
  baseBranch?: string
}

/**
 * Create an isolated worktree for an agent session.
 *
 * Worktrees live under `~/.operator/worktrees/<project>-<short>` on a new
 * branch `operator/<short>`. If the source folder isn't a git repo, this
 * throws — callers should `inspectRepo` first.
 */
export async function createWorktree(sourceCwd: string): Promise<WorktreeCreateResult> {
  const info = await inspectRepo(sourceCwd)
  if (!info.isRepo || !info.root) {
    throw new Error('Not a git repository')
  }

  // Ensure the repo has at least one commit; `git worktree add` from an empty
  // repo fails in confusing ways.
  try {
    await git(info.root, 'rev-parse', 'HEAD')
  } catch {
    throw new Error('Repository has no commits yet — make an initial commit before using worktrees')
  }

  const projectName = basename(info.root)
  const short = Math.random().toString(36).slice(2, 8)
  const branch = `operator/${short}`
  const path = join(WORKTREE_ROOT, `${projectName}-${short}`)

  await fs.mkdir(WORKTREE_ROOT, { recursive: true })

  // Branch from current HEAD of the source repo.
  await git(info.root, 'worktree', 'add', '-b', branch, path, 'HEAD')

  return { path, branch, baseBranch: info.branch }
}

export interface WorktreeStatus {
  branch?: string
  /** Count of changed paths (porcelain output lines). */
  changes: number
  /** Whether the worktree directory exists and is a valid git checkout. */
  valid: boolean
}

export async function worktreeStatus(path: string): Promise<WorktreeStatus> {
  try {
    const [porcelain, branch] = await Promise.all([
      git(path, 'status', '--porcelain'),
      git(path, 'rev-parse', '--abbrev-ref', 'HEAD').catch(() => ''),
    ])
    const changes = porcelain ? porcelain.split('\n').filter(Boolean).length : 0
    return { valid: true, branch: branch || undefined, changes }
  } catch {
    return { valid: false, changes: 0 }
  }
}

export interface FileChange {
  path: string
  status: string // 2-char porcelain code, e.g. " M", "??", "A "
  added: number
  removed: number
}

export interface WorktreeDiff {
  branch?: string
  files: FileChange[]
  /** Unified diff text (no color). Untracked files included with synthetic +lines. */
  diff: string
}

export async function worktreeDiff(path: string): Promise<WorktreeDiff> {
  let branch: string | undefined
  try {
    branch = await git(path, 'rev-parse', '--abbrev-ref', 'HEAD')
  } catch { /* detached */ }

  // Porcelain status to identify per-file state including untracked.
  const porcelain = await git(path, 'status', '--porcelain').catch(() => '')
  const files: FileChange[] = []
  const untracked: string[] = []
  for (const line of porcelain.split('\n')) {
    if (!line) continue
    const status = line.slice(0, 2)
    const file = line.slice(3).trim()
    if (status === '??') untracked.push(file)
    files.push({ path: file, status, added: 0, removed: 0 })
  }

  // Per-file +/- line counts via numstat (tracked changes only).
  const numstat = await git(path, 'diff', 'HEAD', '--numstat').catch(() => '')
  for (const line of numstat.split('\n')) {
    if (!line) continue
    const [add, rem, p] = line.split('\t')
    const entry = files.find((f) => f.path === p)
    if (entry) {
      entry.added = Number(add) || 0
      entry.removed = Number(rem) || 0
    }
  }

  // Full unified diff for tracked changes.
  let diff = await git(path, 'diff', 'HEAD', '--no-color').catch(() => '')

  // Append synthetic +diffs for untracked files so the user sees new file contents too.
  for (const u of untracked) {
    try {
      const content = await git(path, 'show', `:0:${u}`).catch(async () => {
        // File isn't staged; read from disk.
        const buf = await fs.readFile(`${path}/${u}`, 'utf-8').catch(() => '')
        return buf
      })
      const lines = content.split('\n').length
      const header = `diff --git a/${u} b/${u}\nnew file\n--- /dev/null\n+++ b/${u}\n@@ -0,0 +1,${lines} @@\n`
      const body = content.split('\n').map((l) => '+' + l).join('\n')
      diff += (diff ? '\n' : '') + header + body
      const entry = files.find((f) => f.path === u)
      if (entry) entry.added = lines
    } catch { /* ignore */ }
  }

  return { branch, files, diff }
}

/**
 * Remove a worktree. The branch is left intact — the user may have work
 * to merge or review. Caller is responsible for any branch cleanup.
 */
export async function removeWorktree(path: string, sourceRoot: string): Promise<void> {
  try {
    await git(sourceRoot, 'worktree', 'remove', path)
  } catch {
    // Force-remove if the directory was deleted out from under git
    await git(sourceRoot, 'worktree', 'remove', '--force', path).catch(() => undefined)
  }
}

/**
 * Stage all changes and commit. Returns the new commit SHA on success.
 */
export async function commitAll(path: string, message: string): Promise<string> {
  await git(path, 'add', '-A')
  // Check whether there's anything to commit; if not, return the current HEAD.
  const status = await git(path, 'status', '--porcelain').catch(() => '')
  if (!status) {
    return await git(path, 'rev-parse', 'HEAD')
  }
  await git(path, 'commit', '-m', message)
  return await git(path, 'rev-parse', 'HEAD')
}

export interface MergeResult {
  ok: boolean
  /** Diagnostic message — populated on failure (e.g. conflicts). */
  message?: string
}

/**
 * Fast-forward or no-ff merge the worktree branch into the base branch in the
 * source repo. The worktree is removed on success. On failure (conflicts, etc.)
 * the worktree and branch are left intact for the user to fix.
 */
export async function mergeBranch(
  worktreePath: string,
  sourceRoot: string,
  branch: string,
  baseBranch: string,
): Promise<MergeResult> {
  // Refuse if the source repo's working tree is dirty — merging would mix changes.
  try {
    const dirty = await git(sourceRoot, 'status', '--porcelain')
    if (dirty) {
      return { ok: false, message: 'Source repo has uncommitted changes — commit or stash before merging.' }
    }
  } catch (e) {
    return { ok: false, message: (e as Error).message }
  }

  // Switch source repo to base branch if needed.
  try {
    const current = await git(sourceRoot, 'rev-parse', '--abbrev-ref', 'HEAD')
    if (current !== baseBranch) {
      await git(sourceRoot, 'checkout', baseBranch)
    }
  } catch (e) {
    return { ok: false, message: `Could not switch to ${baseBranch}: ${(e as Error).message}` }
  }

  // Merge the worktree branch.
  try {
    await git(sourceRoot, 'merge', '--no-ff', '-m', `Merge ${branch}`, branch)
  } catch (e) {
    // Conflicts or other merge failure — abort and bail.
    await git(sourceRoot, 'merge', '--abort').catch(() => undefined)
    return { ok: false, message: `Merge failed: ${(e as Error).message}` }
  }

  // Success — clean up the worktree.
  await removeWorktree(worktreePath, sourceRoot)
  return { ok: true }
}

/**
 * Delete the worktree's branch outright. Removes the worktree first so git
 * lets us drop the branch ref. Use after a session ends and the work is junk.
 */
export async function discardBranch(
  worktreePath: string,
  sourceRoot: string,
  branch: string,
): Promise<void> {
  await removeWorktree(worktreePath, sourceRoot)
  await git(sourceRoot, 'branch', '-D', branch).catch(() => undefined)
}

File: tauri-spike/src-tauri/src/worktree.rs
// Git worktree operations (port of worktree.ts). Each agent session can run in
// an isolated worktree under ~/.operator/worktrees, with in-app diff/merge/discard.

use std::path::PathBuf;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

use serde::Serialize;

fn worktree_root() -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_default();
    PathBuf::from(home).join(".operator").join("worktrees")
}

fn git(cwd: &str, args: &[&str]) -> Result<String, String> {
    let out = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .map_err(|e| e.to_string())?;
    if out.status.success() {
        Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
    } else {
        Err(String::from_utf8_lossy(&out.stderr).trim().to_string())
    }
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RepoInfo {
    pub is_repo: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub root: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
}

pub fn inspect_repo(cwd: &str) -> RepoInfo {
    match git(cwd, &["rev-parse", "--show-toplevel"]) {
        Ok(root) => {
            let branch = git(cwd, &["rev-parse", "--abbrev-ref", "HEAD"])
                .ok()
                .filter(|b| !b.is_empty() && b != "HEAD");
            RepoInfo { is_repo: true, root: Some(root), branch }
        }
        Err(_) => RepoInfo { is_repo: false, root: None, branch: None },
    }
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeCreateResult {
    pub path: String,
    pub branch: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_branch: Option<String>,
}

fn short_id() -> String {
    let nanos = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos();
    format!("{:x}", nanos & 0xffffff)
}

pub fn create_worktree(source_cwd: &str) -> Result<WorktreeCreateResult, String> {
    let info = inspect_repo(source_cwd);
    let root = match (info.is_repo, info.root) {
        (true, Some(r)) => r,
        _ => return Err("Not a git repository".into()),
    };
    git(&root, &["rev-parse", "HEAD"])
        .map_err(|_| "Repository has no commits yet — make an initial commit before using worktrees".to_string())?;

    let project = root.rsplit('/').next().unwrap_or("project").to_string();
    let short = short_id();
    let branch = format!("operator/{short}");
    let path = worktree_root().join(format!("{project}-{short}"));
    let path_str = path.to_string_lossy().to_string();

    std::fs::create_dir_all(worktree_root()).map_err(|e| e.to_string())?;
    git(&root, &["worktree", "add", "-b", &branch, &path_str, "HEAD"])?;

    Ok(WorktreeCreateResult { path: path_str, branch, base_branch: info.branch })
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeStatus {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
    pub changes: usize,
    pub valid: bool,
}

pub fn worktree_status(path: &str) -> WorktreeStatus {
    match git(path, &["status", "--porcelain"]) {
        Ok(porcelain) => {
            let branch = git(path, &["rev-parse", "--abbrev-ref", "HEAD"]).ok().filter(|b| !b.is_empty());
            let changes = porcelain.lines().filter(|l| !l.is_empty()).count();
            WorktreeStatus { valid: true, branch, changes }
        }
        Err(_) => WorktreeStatus { valid: false, branch: None, changes: 0 },
    }
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FileChange {
    pub path: String,
    pub status: String,
    pub added: u32,
    pub removed: u32,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeDiff {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch: Option<String>,
    pub files: Vec<FileChange>,
    pub diff: String,
}

pub fn worktree_diff(path: &str) -> WorktreeDiff {
    let branch = git(path, &["rev-parse", "--abbrev-ref", "HEAD"]).ok();
    let porcelain = git(path, &["status", "--porcelain"]).unwrap_or_default();

    let mut files: Vec<FileChange> = vec![];
    let mut untracked: Vec<String> = vec![];
    for line in porcelain.lines() {
        if line.is_empty() {
            continue;
        }
        let status = line.chars().take(2).collect::<String>();
        let file = line[3..].trim().to_string();
        if status == "??" {
            untracked.push(file.clone());
        }
        files.push(FileChange { path: file, status, added: 0, removed: 0 });
    }

    if let Ok(numstat) = git(path, &["diff", "HEAD", "--numstat"]) {
        for line in numstat.lines() {
            let cols: Vec<&str> = line.split('\t').collect();
            if cols.len() == 3 {
                if let Some(entry) = files.iter_mut().find(|f| f.path == cols[2]) {
                    entry.added = cols[0].parse().unwrap_or(0);
                    entry.removed = cols[1].parse().unwrap_or(0);
                }
            }
        }
    }

    let mut diff = git(path, &["diff", "HEAD", "--no-color"]).unwrap_or_default();

    // Synthetic +diff for untracked files so new file contents show too.
    for u in &untracked {
        let content = std::fs::read_to_string(format!("{path}/{u}")).unwrap_or_default();
        let lines = content.lines().count().max(1);
        let header = format!("diff --git a/{u} b/{u}\nnew file\n--- /dev/null\n+++ b/{u}\n@@ -0,0 +1,{lines} @@\n");
        let body = content.lines().map(|l| format!("+{l}")).collect::<Vec<_>>().join("\n");
        if !diff.is_empty() {
            diff.push('\n');
        }
        diff.push_str(&header);
        diff.push_str(&body);
        if let Some(entry) = files.iter_mut().find(|f| &f.path == u) {
            entry.added = lines as u32;
        }
    }

    WorktreeDiff { branch, files, diff }
}

pub fn remove_worktree(path: &str, source_root: &str) -> Result<(), String> {
    if git(source_root, &["worktree", "remove", path]).is_err() {
        git(source_root, &["worktree", "remove", "--force", path])?;
    }
    Ok(())
}

pub fn commit_all(path: &str, message: &str) -> Result<String, String> {
    git(path, &["add", "-A"])?;
    let status = git(path, &["status", "--porcelain"]).unwrap_or_default();
    if status.is_empty() {
        return git(path, &["rev-parse", "HEAD"]);
    }
    git(path, &["commit", "-m", message])?;
    git(path, &["rev-parse", "HEAD"])
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MergeResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

pub fn merge_branch(worktree_path: &str, source_root: &str, branch: &str, base_branch: &str) -> MergeResult {
    match git(source_root, &["status", "--porcelain"]) {
        Ok(dirty) if !dirty.is_empty() => {
            return MergeResult { ok: false, message: Some("Source repo has uncommitted changes — commit or stash before merging.".into()) }
        }
        Err(e) => return MergeResult { ok: false, message: Some(e) },
        _ => {}
    }
    if let Ok(current) = git(source_root, &["rev-parse", "--abbrev-ref", "HEAD"]) {
        if current != base_branch {
            if let Err(e) = git(source_root, &["checkout", base_branch]) {
                return MergeResult { ok: false, message: Some(format!("Could not switch to {base_branch}: {e}")) };
            }
        }
    }
    if let Err(e) = git(source_root, &["merge", "--no-ff", "-m", &format!("Merge {branch}"), branch]) {
        let _ = git(source_root, &["merge", "--abort"]);
        return MergeResult { ok: false, message: Some(format!("Merge failed: {e}")) };
    }
    let _ = remove_worktree(worktree_path, source_root);
    MergeResult { ok: true, message: None }
}

pub fn discard_branch(worktree_path: &str, source_root: &str, branch: &str) -> Result<(), String> {
    let _ = remove_worktree(worktree_path, source_root);
    let _ = git(source_root, &["branch", "-D", branch]);
    Ok(())
}
